### DISCUSSION TITLE
[Mastermind (Быки-коровы)] Выделение действий (19.04.21)

### DISCUSSION THREAD

**Discussion by citerm** [nergal-perm/tg-core#37] [Идеи]
На встрече 19.04 обсудили:

- В системе всегда должен существовать свободный стол
- Если игрок присоединился к столу, то система создает новый свободный стол с такой же игрой
- Стол имеет тип игры, правила которой он реализует
- Порядок игроков определяется в момент начала игры
- Действующими лицами за столом всегда являются игроки (игрока может представлять бот)
- Подумать, чтобы при выходе игрока из игры (добровольном выходе из-за стола), другие игроки могли продолжить игру без него (или заменить его ботом) - игрок всегда может закончить игру для себя
- Выделили 3 этапа реализации функционала игроков: 1) бот только загадывает число; 2) 2 игрока загадывают числа друг другу и потом отгадывают; 3) игрок и бот загадывают числа друг другу и потом отгадывают (этапы 2 и 3 могут поменяться местами - то есть мультиплеер после реализации соответствующих правил)
- Подумать над тем, чтобы не хранить состояние стола, а для восстановления стола использовать новый пустой стол и проигрывать в нем сохраненную последовательность действий, которая приводит стол в состояние до постановки игры на паузу

Действия:

- присоединиться

1. пользователь присоединяется к столу 
2. игра создает игрока за столом связанного с пользователем
3. игрок может: начать_игру

- начать_игру

1. игрок начинает игру
2. игра создает бота за столом
3. игрок НЕ может совершать действий
4. бот может: загадать_число

- загадать_число

1. бот загадывает число
2. игра фиксирует число-цель для игрока
3. бот НЕ может совершать действий
4. игрок может: сделать_попытку

- сделать_попытку

1. игрок делает попытку
2. игра игра фиксирует попытку
3. игра проверяет попытку: если число-цель отгадано, то игрок НЕ может совершать действий; если число-цель НЕ отгадано, то игрок может: сделать_попытку

План на следующую встречу:

- попробовать что-то написать
- попробовать определить алгоритм действий в ответ на действие игрока в каждой подсистеме (что делает действие, какие состояния поменяются, какие способности поменяются, какие действия будут доступны)

**Comment by nergal-perm** [comment]
Очень четко в этом описании прослеживаются взаимосвязи между тремя источниками данных:
1. Выполнение действия приводит к определенному изменению состояния
2. Текущее состояние определяет доступность действий для разных игроков

**Reply by citerm** [reply]
я специально старался не забывать про все эффекты действий, чтобы можно было выделить методологию описания, которое приводит к непротиворечивой бизнес логике.
